lively.collab/comments/commentMorph.js

```diff
@@ -0,0 +1,360 @@
+    this.comment = comment;
+    this.referenceMorph = referenceMorph;
+    this.initializeUI();
+    this.initializeCommentIndicator();
+    this.setDate();
+    this.setUser();
+  }
+
+  setDate () {
+    const [date, time] = new Date(this.comment.timestamp).toLocaleString('de-DE', { hour12: false }).split(', ');
+    this.ui.dateLabel.textString = date + ' ' + time;
+  }
+
+  setUser () {
+    let username = this.comment.username;
+    if (username.startsWith('guest')) {
+      username = 'guest';
+    }
+    this.ui.usernameLabel.textString = username;
+  }
+
+  // -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-
+  // ui
+  // -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-
+  initializeUI () {
+    this.reset();
+
+    this.ui.commentTextField.textString = this.comment.text;
+    this.setDefaultUI();
+  }
+
+  setDefaultUI () {
+    Icon.setIcon(this.ui.editSaveButton, 'pencil-alt');
+    Icon.setIcon(this.ui.resolveButton, this.comment.resolved ? 'undo-alt' : 'check');
+
+    // Edit/Save Icon widths are off by one, revert the additional padding from setEditUI
+    if (this.isInEditMode) this.ui.editSaveButton.padding.width -= 1;
+
+    this.isInEditMode = false;
+
+    this.enabledButtonAppearance();
+    this.ui.editSaveButton.tooltip = 'Edit Comment';
+
+    this.ui.commentTextField.readOnly = true;
+    this.ui.commentTextField.fill = Color.rgb(251, 252, 252);
+    this.ui.commentTextField.borderStyle = 'none';
+  }
+
+  enabledButtonAppearance () {
+    const buttonColor = Color.rgb(127, 140, 141);
+
+    this.ui.resolveButton.fontColor = buttonColor;
+    this.ui.resolveButton.nativeCursor = 'pointer';
+    this.ui.resolveButton.tooltip = this.comment.isResolved() ? 'Unresolve Comment' : 'Resolve Comment';
+
+    this.ui.deleteButton.fontColor = buttonColor;
+    this.ui.deleteButton.nativeCursor = 'pointer';
+    this.ui.deleteButton.tooltip = 'Delete Comment';
+  }
+
+  disabledButtonAppearance () {
+    const deactivatedButtonColor = Color.rgba(127, 140, 141, 0.4);
+
+    this.ui.resolveButton.fontColor = deactivatedButtonColor;
+    this.ui.resolveButton.nativeCursor = 'default';
+    this.ui.resolveButton.tooltip = 'Save comment to be able to resolve it';
+
+    this.ui.deleteButton.fontColor = deactivatedButtonColor;
+    this.ui.deleteButton.nativeCursor = 'default';
+    this.ui.deleteButton.tooltip = 'Save comment to be able to delete it';
+  }
+
+  setEditUI () {
+    Icon.setIcon(this.ui.editSaveButton, 'save');
+
+    // Edit/Save Icon widths are off by one, will be reverted in setDefaultUI
+    if (!this.isInEditMode) this.ui.editSaveButton.padding.width += 1;
+
+    this.isInEditMode = true;
+
+    this.disabledButtonAppearance();
+    this.ui.editSaveButton.tooltip = 'Save Comment';
+
+    this.ui.commentTextField.readOnly = false;
+    this.ui.commentTextField.fill = Color.white;
+    this.ui.commentTextField.borderStyle = 'solid';
+    this.ui.commentTextField.focus();
+  }
```

Comment:
We are initializing the ui via code because we want to set it do default in different cases. It is a master component but when a button is clicked different icons are removed and altered so to set it back to default we are doing it here in code. How could we do this via master components to get this behavior?

Reply:
I am picturing different master components for each situation: a default style master component, and other ones for the other cases respectively. I don't really see you altering the submorph collections. And even that can be handled with showing or hiding certain morphs in the different master components.